### ISSUE TITLE
Fix S2583 FP: Remainder calculation in if statement

### ISSUE THREAD

**Issue by thef1chesser** [SonarSource/sonar-dotnet#8108]
### Description

I'm getting a false positive S2583 message on a remainder calculation in an if statement.
The part: “changesPushed % 1000 == 0” is marked with a S2583 Change this condition so that it does not always evaluate to ‘False’. Some code paths are unreachable.

![image](https://github.com/SonarSource/sonar-dotnet/assets/22844191/17e789d2-80d5-4fd9-8b31-3d240c4de414)

### Repro steps

```
// Add a log every 1000 changes
if (changesPushed % 1000 == 0 && changesPushed != 0)
{
    // Add log
}
```

### Expected behavior

Warning should not show

### Actual behavior

Warning shows

### Known workarounds

It's a warning not an error so can just continue.

### Related information

Visual Studio team has forwarded me to SonarSource.
SonarLint version 7.3.0.77872
Visual Studio version 17.7.3 & 17.7.4
dotnet version - project is .Net 6.0
Operating System - Windows 10.0

**Comment by antonioaversa** [comment]
Hi @thef1chesser, thanks for reporting the issue.

The rule [S2583](https://sonarsource.github.io/rspec/#/rspec/S2583/csharp) is one of those rules performing symbolic execution to detect issues: it analyses the flow of operations and determines solution spaces for the variables in the state of the program.
So, in order to conclude whether the scenario you are showing is indeed a False Positive, I would need to know more about the context and the flow of operation leading to this `if` statement. 

For example, in the following example, we don't report any issue on the condition of the `if` statement, despite it being always `false`, because our symbolic execution for .NET is currently only mono-procedural (i.e. doesn't look into `F`).
```cs
class C
{
    int F() => 1;

    void Method()
    {
        var i = F();
        if (i % 1000 == 0 && i != 0) { } // False Negative 
    }
}
```
In the following case, however, we report an issue:
```cs
class C
{
    int F() => 0;

    void Method()
    {
        var i = 1;
        if (i % 1000 == 0 && i != 0) { } // Noncompliant
        //  ^^^^^^^^^^^^^
    }
}
```

From the tooltip shown by Visual Studio, I see that `changesPushed` is a local variable: where and how it is declared, what is its initial value, and how its value is kept across multiple executions of the method performing changes?

Best regards,
Antonio

**Comment by thef1chesser** [comment]
changesPushed is a local variable starting with the value 0, it then goes into a foreach loop where it is incremented right before the if statement as can be seen in the screenshot.

**Comment by antonioaversa** [comment]
@thef1chesser Thanks for the clarification. I can confirm the False Positive.
The following code is the minimal example that reproduces the issue for me:
```cs
class C
{
    void Method()
    {
        var i = 0;
        foreach (var _ in Enumerable.Range(0, 10))
        {
            i++;
            if (i % 1000 == 0 && i != 0) { } // False Positive
            //  ^^^^^^^^^^^^^
        }
    }
}
```

We are seeing similar issues with loop unrolling (that is the technique used to run symbolic executions on loops) in multiple rules, and we plan to run a dedicated hardening sprint on the subject later this year. I am going to add this issue to our backlog so that it will be taken into account in our planning.

For the time being, and as a temporary solution:
- If you are using SonarLint in connected mode: you can mark the issue as a false-positive in the UI of SonarQube or SonarCloud. The rule can also be disabled in your Quality Profile if gets particularly noisy in your codebase.
- If you are using SonarLint in Standalone Mode: you can use `#pragma warning disable` and `#pragma warning restore` to locally disable the rule in the source, or the `.editorconfig` files to disable or change the severity of the rule.

Apologies for the inconvenience.

**Comment by Tim-Pohlmann** [comment]
I close this as a duplicate of #8028 to track all issues caused by loop counters not being fully explored in one place.
